### DISCUSSION TITLE
How to pipe commands to ImageMagick when it has been spawned by another program?

### DISCUSSION THREAD

**Discussion by danielwlewis** [ImageMagick/ImageMagick#6261] [Help]
### ImageMagick version

7.0.10-31

### Operating system, version and so on

Windows 10 Professional

### Description

I have a C program that builds a text file called "cmds.txt", containing a list of ImageMagick commands to create thumbnails from fullsize images that are in a variety of folders, as in:

```
red\file1.jpg[80x100] -write purple\thumb1.jpg -delete 0
blue\file1.jpg[80x100] -write orange\thumb1.jpg -delete 0
...
green\file1.jpg[80x100] -write black\thumb1.jpg -delete 0
```

The C program then spawns ImageMagick with a script command line option to process the list, as in:

```
    char *args[4] ;

    args[0] = "magick" ;
    args[1] = "-script" ;
    args[2] = "cmds.txt" ;
    args[3] = NULL ;

   // Spawn the following command: "magick -script cmds.txt"
    _spawnvp(_P_NOWAIT, args[0], (const char * const *) args) ;
```
I'd like to change it so that the C program spawns magick, which then waits for commands via a pipe.  If I understand correctly, I will need to add one last command ("-exit") to tell magick to terminate after all the images have been processed.

My question: What should the command line options of magick be when it is spawned so that it knows to get the commands from the pipe?

Thanks!
Dan

**Comment by snibgo** [comment]
> I'd like to change it so that the C program spawns magick, which then waits for commands via a pipe.

I don't think that is possible with `magick.exe`. It is possible with `magick-script.exe`.

Check whether your IM installation directory contains `magick-script.exe`. If it doesn't, then create it as a copy of`magick.exe`.

The program `magick-script.exe` can take all its arguments from `stdin`, by invoking it with one argument, `-`, thus:
```
magick-script.exe -
```
You can then type an IM command, ending in `-exit`. I suggest you use explicit `-read` and `-write` options.

You can send commands to `magick-script.exe` via anonymous pipes, like this (Windows syntax):
```
echo -read toes.png -resize 200% -write x.png -exit | magick-script -
```
You can also use named pipes. These are available in Windows, but not through the usual CMD CLI. They are available through Cygwin bash. See [Client-server](http://im.snibgo.com/cliserv.htm) for examples. That page also includes a program similar to `magick-script.exe` but taking any number of independent IM commands.
